### PR TITLE
Add Nudge for mentoring

### PR DESCRIPTION
**Pull request by iHiD**
This will run daily and send an email to everyone who has submitted a non-hello-world practice solution that's not been mentored. People only ever get one notification.

Note: This needs an entry adding the schedule - but I'll run it manually the first time, so I've not done that for now.

### CODE REVIEW

File: db/migrate/20211130133210_add_index_for_nudge_job.rb
Comment (iHiD) [inline db/migrate/20211130133210_add_index_for_nudge_job.rb]: These need removing before I merge.

File: db/migrate/20211130133210_add_index_for_nudge_job.rb
Comment (iHiD) [inline db/migrate/20211130133210_add_index_for_nudge_job.rb]: This needs the algorthm/lock thing adding and the name removing.

File: app/jobs/nudge_users_to_request_mentoring_job.rb
Comment (ErikSchierboom) [inline app/jobs/nudge_users_to_request_mentoring_job.rb:13]: Good idea

File: app/jobs/nudge_users_to_request_mentoring_job.rb
Comment (ErikSchierboom) [inline app/jobs/nudge_users_to_request_mentoring_job.rb:8]: With all the comment, this query is strangely pleasing to read :)

File: app/views/mailers/notifications_mailer/nudge_to_request_mentoring.html.haml
Comment (ErikSchierboom) [inline app/views/mailers/notifications_mailer/nudge_to_request_mentoring.html.haml:2]: At first I thought: "a few exercise", but then I realized that students will all have done `hello-world` :) (although you can't request mentoring on that)

File: app/views/mailers/notifications_mailer/nudge_to_request_mentoring.html.haml
Comment (ErikSchierboom) [inline app/views/mailers/notifications_mailer/nudge_to_request_mentoring.html.haml:4]: Is it `levelling` in US English or `leveling`? I've done a quick search and found:

https://en.wikipedia.org/wiki/Levelling#:~:text=Levelling%20or%20leveling%20(American%20English,points%20relative%20to%20a%20datum.&text=It%20is%20also%20known%20as%20spirit%20levelling%20and%20differential%20levelling.
https://ell.stackexchange.com/questions/146181/in-us-english-which-is-the-appropriate-spelling-leveled-a-look-or-levelled-a-l

It _looks_ like it should be `leveling`, but 🤷

File: app/views/mailers/notifications_mailer/nudge_to_request_mentoring.text.haml
Comment (ErikSchierboom) [inline app/views/mailers/notifications_mailer/nudge_to_request_mentoring.text.haml]:
```suggestion
Receiving feedback from our mentors is an amazing way of discovering your knowledge gaps and levelling up your skills. It's one of Exercism's most valuable features, and like all of Exercism, it's absolutely free!
```

File: app/views/mailers/notifications_mailer/nudge_to_request_mentoring.text.haml
Comment (ErikSchierboom) [inline app/views/mailers/notifications_mailer/nudge_to_request_mentoring.text.haml]:
```suggestion
Choose a solution that you feel has the most room for improvement and request a mentor to take a look at it. They should get back to you within a day or so with some ideas!
```

File: app/views/mailers/notifications_mailer/nudge_to_request_mentoring.text.haml
Comment (ErikSchierboom) [inline app/views/mailers/notifications_mailer/nudge_to_request_mentoring.text.haml]: Maybe:

```suggestion
Choose a solution at: #{@notification.url}
```

File: db/migrate/20211130133210_add_index_for_nudge_job.rb
Comment (ErikSchierboom) [inline db/migrate/20211130133210_add_index_for_nudge_job.rb]: Will this temporarily lock the table? I'm not sure how mysql does this.

File: test/jobs/nudge_users_to_request_mentoring_job_test.rb
Comment (ErikSchierboom) [inline test/jobs/nudge_users_to_request_mentoring_job_test.rb:14]: Should there be a call to `NudgeUsersToRequestMentoringJob.perform_now` after this to verify that the notification is not sent?

```ruby
User::Notification::Create.expects(:call).never
NudgeUsersToRequestMentoringJob.perform_now
```

If correct, you probably want to do this for the other exception cases below too.

File: config/locales/notifications/en.yml
Comment (ErikSchierboom) [inline config/locales/notifications/en.yml:32]: Maybe be slightly more specific on what we expect?

```suggestion
        Learning with our mentors is an amazing way to level up your knowledge, and it's 100% free. Choose one of your solutions to request mentoring on.
```

File: app/jobs/nudge_users_to_request_mentoring_job.rb
Comment (ErikSchierboom) [inline app/jobs/nudge_users_to_request_mentoring_job.rb:34]: Does `.last` by default return the latest track? I'm asking because there is no explicit sorting here.

**Review by ErikSchierboom — APPROVED** [review]

File: app/jobs/nudge_users_to_request_mentoring_job.rb
Comment (iHiD) [reply app/jobs/nudge_users_to_request_mentoring_job.rb:34]: It will return the last practice solution, which will be default ordered by the primary key, so it'll be the last submitted.

File: app/views/mailers/notifications_mailer/nudge_to_request_mentoring.html.haml
Comment (iHiD) [reply app/views/mailers/notifications_mailer/nudge_to_request_mentoring.html.haml:2]: Yes - I wondered if we should hold off and only do it when someone has done three exercises, but it made the query too heavy :)

File: db/migrate/20211130133210_add_index_for_nudge_job.rb
Comment (iHiD) [reply db/migrate/20211130133210_add_index_for_nudge_job.rb]: Not if you use the algorithm+lock commands as per https://github.com/exercism/website/pull/2197/files#diff-e85712804207d693a50b13df15695b213427e6f51c2eab47a534762de03a3230R3 - Mysql now supports non-blocking table modifications if those are specified.

File: test/jobs/nudge_users_to_request_mentoring_job_test.rb
Comment (iHiD) [reply test/jobs/nudge_users_to_request_mentoring_job_test.rb:14]: Because L35 specifies uses `expects` on `User::Notification::Create`, if only allows calls with the exact params. If that gets called with any of the other users, it explodes. So this is actually tested, but I'll add a comment below to make that clear.

File: test/jobs/nudge_users_to_request_mentoring_job_test.rb
Comment (iHiD) [inline test/jobs/nudge_users_to_request_mentoring_job_test.rb:36]:
```suggestion
    # Only expect call with the initial correct user. This will raise if called with any other user.
    User::Notification::Create.expects(:call).with(user, :nudge_to_request_mentoring, track: ruby)
```

File: config/locales/notifications/en.yml
Comment (iHiD) [reply config/locales/notifications/en.yml:32]: I'm going to leave this as it is, purely so it doesn't roll over onto another line!

File: app/views/mailers/notifications_mailer/nudge_to_request_mentoring.html.haml
Comment (iHiD) [reply app/views/mailers/notifications_mailer/nudge_to_request_mentoring.html.haml:4]: I'm going to leave this one I think. One `l` just looks too wrong to me!!

File: db/migrate/20211130133210_add_index_for_nudge_job.rb
Comment (ErikSchierboom) [reply db/migrate/20211130133210_add_index_for_nudge_job.rb]: Cool

File: test/jobs/nudge_users_to_request_mentoring_job_test.rb
Comment (ErikSchierboom) [reply test/jobs/nudge_users_to_request_mentoring_job_test.rb:14]: Ah okay. Thanks

**Pull request merged**
